### PR TITLE
Implemented GitHub Actions cache example

### PR DESCRIPTION
**Pull request by leozz37**
### GitHub Actions Cache

Implemented example of GitHub Actions Cache usage, to run CI faster.

### CODE REVIEW

**Comment by TheLartians** [comment]
Hey, great idea with the actions cache, that can really help for faster CI workflows! Even better would be ccache support, which would allow us to cache the built binaries as well.

Perhaps instead of a dedicated section in the readme, this would deserve its own [wiki page](https://github.com/TheLartians/CPM.cmake/wiki) and readme could link there?

**Comment by dominiklohmann** [comment]
For ccache, I have written this before, which works for pretty much every CI workflow.

I'm pretty sure this needs some adjustment to work on Windows. We only run Linux/macOS CI at work.

```yaml
[...]
    env: 
      CCACHE_HASH_DIR: true
      CCACHE_UNIFY: true
      CCACHE_SLOPPINESS: 'file_macro,time_macros'
      CCACHE_DIR: '${{ github.workspace }}/.ccache'
[...]
      - name: Configure Cache
        run: |
          # Export slug variables for the cache action (requires gsed on macOS).
          slug_ref() {
            echo "$1" |
              tr "[:upper:]" "[:lower:]" |
              ${{ matrix.os.tag == 'macOS' && 'gsed' || 'sed' }} -r 's#refs/[^\/]*/##;s/[~\^]+//g;s/[^a-zA-Z0-9.]+/-/g;s/^-+\|-+$//g;s/^-*//;s/-*$//' |
              cut -c1-63
          }
          # Identify OS, Compiler, and CMake Configuration based on the configuration matrix.
          echo ::set-env name=CACHE_PREFIX::"${{ matrix.os.tag }}-${{ matrix.os.compiler }}-${{ matrix.configure.tag }"
          echo ::set-env name=CACHE_MAIN_BRANCH::"master"
          echo ::set-env name=CACHE_REF_SLUG::"$(slug_ref "$GITHUB_REF")"
          echo ::set-env name=CACHE_HEAD_REF_SLUG::"$(slug_ref "$GITHUB_HEAD_REF")"
          echo ::set-env name=CACHE_BASE_REF_SLUG::"$(slug_ref "$GITHUB_BASE_REF")"
[...]
      # For 'pull_request' events we want to take the latest build on the PR
      # branch, or if that fails the latest build from the branch we're merging
      # into.
      - name: Fetch ccache Cache (Pull Request)
        if: github.event_name == 'pull_request'
        uses: actions/cache@v2
        with:
          path: .ccache
          key: ccache-${{ env.CACHE_PREFIX }}-${{ env.CACHE_HEAD_REF_SLUG }}-${{ github.sha }}
          restore-keys: |
            ccache-${{ env.CACHE_PREFIX }}-${{ env.CACHE_HEAD_REF_SLUG }}
            ccache-${{ env.CACHE_PREFIX }}}-${{ env.CACHE_BASE_REF_SLUG }}
            ccache-${{ env.CACHE_PREFIX }}
      # For 'push' events we want to take the latest build on the branch we pushed to.
      - name: Fetch ccache Cache (Push)
        if: github.event_name == 'push'
        uses: actions/cache@v2
        with:
          path: .ccache
          key: ccache-${{ env.CACHE_PREFIX }}-${{ env.CACHE_REF_SLUG }}-${{ github.sha }}
          restore-keys: |
            ccache-${{ env.CACHE_PREFIX }}-${{ env.CACHE_REF_SLUG }}
            ccache-${{ env.CACHE_PREFIX }}
      # For 'release' events we want to take the latest master build.
      - name: Fetch ccache Cache (Release)
        if: github.event_name == 'release'
        uses: actions/cache@v2
        with:
          path: .ccache
          key: ccache-${{ env.CACHE_PREFIX }}-${{ env.CACHE_MAIN_BRANCH }}-${{ github.sha }}
          restore-keys: |
            ccache-${{ env.CACHE_PREFIX }}-${{ env.CACHE_MAIN_BRANCH }}
            ccache-${{ env.CACHE_PREFIX }}
[...]
      # Do the actual build and configuration
      - name: Configure / Build
[...]
      - name: Show ccache Statistics
        run: |
          # Print statistics counter IDs and corresponding values.
          ccache --show-stats
```

**Comment by TheLartians** [comment]
Thanks for the example @dominiklohmann! May I ask what the advantage is of splitting the configuration for pushes, PRs and release? E.g. [this tutorial](https://cristianadam.eu/20200113/speeding-up-c-plus-plus-github-actions-using-ccache/) has much less configuration code.

**Comment by dominiklohmann** [comment]
The variables `$GITHUB_REF`, `$GITHUB_HEAD_REF`, and `$GITHUB_BASE_REF` have different meanings depending on the event that triggered the workflow run. The split allows for better approximation which cache we want to use. E.g., for a `pull_request` event, we want to use the last cache from within that pull request, falling back to the cache from the branch we're merging into.

**Comment by TheLartians** [comment]
I see, thanks for the explanation!

**Comment by leozz37** [comment]
Hey, I create a new [wiki page](https://github.com/TheLartians/CPM.cmake/wiki/Caching-with-CPM.cmake-and-ccache-on-GitHub-Actions) with both CPM.cmake and ccache example. For ccache I used [the tutoral](https://cristianadam.eu/20200113/speeding-up-c-plus-plus-github-actions-using-ccache/) that Lars talked about.

I think we should put more information about CMake setup to use ccache, since we're talking about GitHub Actions, didn't address about.

I'll be updating my PR to link to the wiki page.

**Review by TheLartians — CHANGES_REQUESTED** [review]
Looks good! Requesting a small change for consistency.

File: README.md [hunk 1]
```diff
@@ -155,6 +155,8 @@ export CPM_SOURCE_CACHE=$HOME/.cache/CPM
 
 Note that passing the variable as a configure option to CMake will always override the value set by the environmental variable.
 
+You can use CPM_SOURCE_CACHE on GitHub Actions workflows [cache](https://github.com/actions/cache) and combine it with ccache, to make your CI faster. See the [wiki](https://github.com/TheLartians/CPM.cmake/wiki/Caching-with-CPM.cmake-and-ccache-on-GitHub-Actions) for more info.
```
Comment (TheLartians) [inline README.md:158]:
```suggestion
You can use `CPM_SOURCE_CACHE` on GitHub Actions workflows [cache](https://github.com/actions/cache) and combine it with ccache, to make your CI faster. See the [wiki](https://github.com/TheLartians/CPM.cmake/wiki/Caching-with-CPM.cmake-and-ccache-on-GitHub-Actions) for more info.
```

**Review by TheLartians — APPROVED** [review]

**Pull request merged**